_summaries/katharopoulos2020lineartransformer_2.md

```diff
@@ -0,0 +1,58 @@
+---
+layout: summary
+title: Summary
+paper: {{katharopoulos2020lineartransformer}}
+author:  sritank
+score:  10
+---
+
+## What is the core idea?
+A new linear transformer model is introduced that uses a kernel based self-attention method that reduces the memory footprint of the model. They also linearize the masking of the attention computation making it run in linear time.
+
+## How is it realized (technically)?
+
+They change the attention mechanism from the traditional softmax function to a dot product feature map. In a traditional transformer, given the queries Q, keys K and values V, the self attention function A_l can be calculated as:
+
+![self attention function](./katharopoulos2020lineartransformer_2a.png)
+
+where,
+
+![softmax kernel](./katharopoulos2020lineartransformer_2b.png)
+
+Instead of a softmax similarity function, the authors propose using a kernel function of a feature representation \phi(x). So the new self attention function is:
+
+![linear self attention function](./katharopoulos2020lineartransformer_2c.png)
+
+The above attention function has computation cost of O(N) where N is the sequence length. It's memory footprint is also lower as the terms can be reused for every query hence we don't need to store the full attention matrix making it's memory complexity O(N). The authors state that for small sequences the below feature map, phi(x), was used and shown to perform on par to the full traditional transformer.
+
+They further change the masking function to span until the i-th position of the query instead of the whole sequence (i.e. all N inputs). They called this ``causal'' masking. They further show how their formulation of a transformer with causal masking can be understood as an RNN.
```

Comment:
I think you should mention the motivation behind why they even want to represent the transformer layer as an RNN. Something like - the linear transformer models allows parallelization during training, but also allows much faster inference.

Reply:
Not sure I understand. I thought transformers were already parallelizable during training. RNNs and LSTM training are not parallelizable because of their sequential nature which was one of their biggest limitations. I guess their motivations for expressing it as RNN is to better understand how a transformer works and what attention does.

Reply:
I was referring to how in the paper they mentioned how their idea of a "Linear Transformer" had the "best of both worlds" because it allows for parallelization during training and much faster inference compared to standard transformers. 

You can ignore the first sentence of my comment - I think I got some stuff mixed up.

Reply:
Got it. I will mention this during the presentation